File: 10_dijkstra/Dijkstra.java
// CN Lab Experiment 10: Dijkstra's Shortest Path Algorithm
// 0 means no direct edge; positive values are edge weights.

import java.util.*;

public class Dijkstra {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        final int INF = 1_000_000_000;

        System.out.print("Enter number of vertices: ");
        int n = scanner.nextInt();

        int[][] graph = new int[n][n]; // Store the adjacency matrix.

        System.out.println("Enter adjacency matrix (0 means no edge):");
        for (int i = 0; i < n; i++) {             // Read each row.
            for (int j = 0; j < n; j++) {         // Read each column.
                graph[i][j] = scanner.nextInt();
            }
        }

        System.out.print("Enter source vertex (0-based): ");
        int source = scanner.nextInt();

        int[] distance = new int[n];              // Shortest known distances.
        boolean[] visited = new boolean[n];       // Finalized vertices.
        Arrays.fill(distance, INF);
        distance[source] = 0;

        for (int count = 0; count < n; count++) { // Finalize vertices one by one.
            int u = -1;
            int best = INF;

            for (int v = 0; v < n; v++) {         // Find nearest unvisited vertex.
                if (!visited[v] && distance[v] < best) {
                    best = distance[v];
                    u = v;
                }
            }

            if (u == -1) break;                   // Remaining vertices are unreachable.

            visited[u] = true;

            for (int v = 0; v < n; v++) {         // Relax all neighbors.
                if (graph[u][v] > 0 && !visited[v]) {
                    int newDistance = distance[u] + graph[u][v];
                    if (newDistance < distance[v]) {
                        distance[v] = newDistance;
                    }
                }
            }
        }

        System.out.println("Shortest distances from vertex " + source);
        for (int v = 0; v < n; v++) {             // Print final distances.
            System.out.println(v + " -> " +
                (distance[v] == INF ? "INF" : distance[v]));
        }

        scanner.close();
    }
}

File: 10_dijkstra/dijkstra.py
# CN Lab Experiment 10: Dijkstra's Shortest Path Algorithm
# Use 0 to mean "no direct edge" and positive values for edge weights.

INF = 10**9

n = int(input("Enter number of vertices: "))
print("Enter adjacency matrix (0 means no edge):")
graph = []
for _ in range(n):
    graph.append(list(map(int, input().split())))  # Read one matrix row.

source = int(input("Enter source vertex (0-based): "))

distance = [INF] * n       # Shortest known distance to each vertex.
visited = [False] * n      # True after a vertex is finalized.
distance[source] = 0

for _ in range(n):         # Repeat until all possible vertices are finalized.
    u = -1
    best = INF

    for v in range(n):     # Find the unvisited vertex with smallest distance.
        if not visited[v] and distance[v] < best:
            best = distance[v]
            u = v

    if u == -1:
        break              # Remaining vertices are unreachable.

    visited[u] = True

    for v in range(n):     # Relax every neighbor of the selected vertex.
        if graph[u][v] > 0 and not visited[v]:
            new_distance = distance[u] + graph[u][v]
            if new_distance < distance[v]:
                distance[v] = new_distance

print("Shortest distances from vertex", source)
for v in range(n):         # Display the final distance to each vertex.
    if distance[v] == INF:
        print(v, "-> INF")
    else:
        print(v, "->", distance[v])
